File: backend/app/seed.py
from __future__ import annotations

import json
from pathlib import Path

from sqlalchemy.orm import Session

from app.core.config import settings
from app.core.security import hash_password
from app.models.models import AdminUser, Resume, SiteSettings, TeamMember

DEFAULT_HOME = {
    "hero": {
        "eyebrow": "[ پروتکل جنسیس ]",
        "line1": "از خلأ",
        "line2_prefix": "از",
        "line2_stroke": "خالی است",
        "line3_prefix": "ما می‌سازیم",
        "line3_accent": "نور",
        "description": "سیستم‌های درهم و مسئله‌های سخت را به معماری شفاف تبدیل می‌کنیم — SaaS چندمستاجره، طراحی دامنه، و هسته‌های پرسرعت.",
    },
    "void": {
        "title": "خلأ",
        "subtitle": "جایی که بیشتر محصول‌ها گیر می‌کنند",
        "items": [
            {
                "label": "۰۱",
                "title": "منطق پراکنده",
                "text": "قواعد کسب‌وکار زیر UI، اسکریپت و دانش شفاهی دفن شده — چیزی برای تغییر امن نیست.",
            },
            {
                "label": "۰۲",
                "title": "چندمستأجره‌ی شکننده",
                "text": "فورک‌های سفارشی برای هر مشتری، به‌جای یک زیربنای واقعی چندمستاجره.",
            },
            {
                "label": "۰۳",
                "title": "سیستم‌های مبهم",
                "text": "بدون مرز، بدون رویداد، بدون مشاهده‌پذیری — فقط آتش‌نشانی.",
            },
        ],
    },
    "crossing": {
        "title": "گذار",
        "subtitle": "از آشوب تا سیستمی که می‌ماند",
        "steps": [
            {"from": "نیازمندی‌های درهم", "to": "مدل دامنه دقیق"},
            {"from": "مونولیت شکننده", "to": "باندد کانتکست"},
            {"from": "حالت خالی", "to": "جریان‌های روشن"},
            {"from": "دیباگ قهرمانانه", "to": "پایپ‌لاین مشاهده‌پذیر"},
        ],
    },
    "capabilities": {
        "title": "توانایی‌ها",
        "subtitle": "آنچه می‌سازیم",
        "cards": [
            {
                "icon": "۰۱ // چندمستاجره",
                "title": "اکوسیستم SaaS",
                "description": "زیرساخت ایزوله، امن و مقیاس‌پذیر — یک بنیاد، قلمروهای بسیار.",
                "span": 2,
            },
            {
                "icon": "۰۲ // DDD",
                "title": "معماری سیستم",
                "description": "نظم روی آشوب. کانتکست‌های خالص و مقیاس‌پذیر.",
                "span": 1,
            },
            {
                "icon": "۰۳ // محصول",
                "title": "ذهنیت محصول",
                "description": "هر خط کد برای نبض واقعی محصول است.",
                "span": 1,
            },
            {
                "icon": "هسته منطق // پایتون و جنگو",
                "title": "منطق کسب‌وکار",
                "description": "منطق پیچیدهٔ چندمستاجره با سرعت آزموده‌شده.",
                "code": "def create_reality(req):\n    if req.void.is_empty:\n        return Architect.build(SaaS())",
                "span": 2,
            },
            {
                "icon": "هسته سیستم // Rust Async",
                "title": "کارایی بالا",
                "description": "انتزاع بدون هزینه؛ جایی که حافظه و سرعت حیاتی‌اند.",
                "code": "async fn illuminate(s: &mut Void) {\n    s.extract_truth().await;\n    Light::from(s)\n}",
                "span": 2,
            },
        ],
    },
    "method": {
        "title": "روش",
        "subtitle": "پروتکل کار ما",
        "steps": [
            {
                "key": "کشف",
                "text": "نقشهٔ درهم‌ریختگی. مصاحبه با دامنه. نام‌گذاری آنچه واقعاً درد می‌کند.",
            },
            {
                "key": "مدل",
                "text": "رسم باندد کانتکست. جدا کردن مسیر نوشتن و خواندن وقتی مقیاس می‌طلبد.",
            },
            {
                "key": "ساخت",
                "text": "سرویس‌های ماژولار — جنگو برای سرعت محصول، Rust وقتی لبه تیز است.",
            },
            {
                "key": "روشنی",
                "text": "مستند، مشاهده‌پذیر، و تحویل سیستمی که تیم مالک آن باشد.",
            },
        ],
    },
    "team": {
        "title": "حلقه",
        "subtitle": "معماران نادیده",
    },
    "contact": {
        "title": "بیایید نور را ببافیم.",
        "subtitle": "آماده‌اید خلأ را پشت سر بگذارید؟ یک سیگنال بفرستید.",
        "button": "ارسال سیگنال ←",
    },
}


def empty_resume(name_en: str, name_fa: str, role_en: str, role_fa: str, email: str = "") -> dict:
    return {
        "personal": {
            "name": {"fa": name_fa or name_en, "en": name_en},
            "tagline": {"fa": role_fa or role_en, "en": role_en},
            "avatar": {"fa": (name_fa or name_en)[:1], "en": name_en[:1]},
            "location": {"fa": "تهران", "en": "Tehran"},
            "phone": {"fa": "", "en": ""},
            "phoneHref": "",
            "email": email,
            "badges": [],
        },
        "contactsSecondary": [],
        "summary": {"fa": "", "en": ""},
        "experiences": [],
        "projects": {"major": [], "small": []},
        "skills": [],
        "education": [],
        "certificates": [],
        "languages": [],
        "footer": {
            "text": {"fa": "Genesis", "en": "Genesis"},
            "backToTop": {"fa": "بازگشت به بالا", "en": "Back to top"},
        },
    }


def seed_if_empty(db: Session) -> None:
    if db.query(AdminUser).first():
        return

    db.add(
        AdminUser(
            username=settings.admin_username,
            password_hash=hash_password(settings.admin_password),
        )
    )

    profile_path = Path(__file__).resolve().parents[2] / "src" / "data" / "profile.json"
    ali_resume: dict = {}
    contacts = {
        "email": "alimahdavinia125@gmail.com",
        "telegram": "https://t.me/mahdaviniaali",
        "linkedin": "https://www.linkedin.com/in/mahdaviniaali/",
        "github": "https://github.com/mahdaviniaali",
    }
    if profile_path.exists():
        raw = json.loads(profile_path.read_text(encoding="utf-8"))
        ali_resume = raw.get("resume", {})
        landing = raw.get("landing", {})
        for c in landing.get("contacts", []):
            t = c.get("type")
            if t in contacts and c.get("href"):
                contacts[t] = c["href"].replace("mailto:", "") if t == "email" else c["href"]

    email_raw = contacts.get("email", "alimahdavinia125@gmail.com")
    email_clean = email_raw.replace("mailto:", "")

    db.add(
        SiteSettings(
            brand_name="جنسیس",
            tagline="از خلأِ خالی است؛ ما نور را می‌سازیم.",
            email=email_clean,
            telegram=contacts.get("telegram", ""),
            linkedin=contacts.get("linkedin", ""),
            github=contacts.get("github", ""),
            home_content=json.dumps(DEFAULT_HOME, ensure_ascii=False),
        )
    )

    ali = TeamMember(
        slug="ali-mahdavinia",
        name_en="Ali Mahdavinia",
        name_fa="علی مهدوی‌نیا",
        role_en="Backend & System Architect",
        role_fa="معمار سیستم و بک‌اند",
        short_bio_en="Leads multi-tenant SaaS and domain-driven backends — turning messy product ideas into durable systems.",
        short_bio_fa="رهبری بک‌اند SaaS چندمستاجره و معماری دامنه — تبدیل ایده‌های مبهم به سیستم‌های پایدار.",
        avatar_url="",
        email=email_clean,
        github=contacts.get("github", ""),
        linkedin=contacts.get("linkedin", ""),
        telegram=contacts.get("telegram", ""),
        sort_order=0,
        is_published=True,
    )
    db.add(ali)
    db.flush()
    db.add(Resume(member_id=ali.id, data_json=json.dumps(ali_resume or empty_resume(
        ali.name_en, ali.name_fa, ali.role_en, ali.role_fa, ali.email
    ), ensure_ascii=False)))

    sara = TeamMember(
        slug="sara-nokhavat",
        name_en="Sara Nokhavat",
        name_fa="سارا نخاوت",
        role_en="Product & Domain Strategist",
        role_fa="استراتژیست محصول و دامنه",
        short_bio_en="Turns ambiguous briefs into crisp requirements and domain language the whole team can ship against.",
        short_bio_fa="بریف‌های مبهم را به نیازمندی و زبان دامنه شفاف برای کل تیم تبدیل می‌کند.",
        sort_order=1,
        is_published=True,
        email="sara@genesis.dev",
    )
    db.add(sara)
    db.flush()
    sara_resume = empty_resume(sara.name_en, sara.name_fa, sara.role_en, sara.role_fa, sara.email)
    sara_resume["summary"] = {
        "en": "Product strategist focused on domain clarity. Bridges founders and engineers so Genesis ships systems that match the real business — not a guess.",
        "fa": "استراتژیست محصول با تمرکز روی شفافیت دامنه. پل بین بنیان‌گذار و مهندس تا سیستم با واقعیت بیزینس هم‌خوان باشد.",
    }
    sara_resume["skills"] = ["Domain Workshops", "User Research", "PRD / Spec Writing", "SaaS Roadmaps", "Stakeholder Alignment"]
    sara_resume["experiences"] = [
        {
            "title": {"en": "Product Strategist", "fa": "استراتژیست محصول"},
            "location": {"en": "Genesis", "fa": "جنسیس"},
            "date": {"en": "2024 — Present", "fa": "۱۴۰۳ — اکنون"},
            "description": {
                "en": "Runs discovery workshops, shapes backlog language, and keeps engineering aligned with revenue-critical flows.",
                "fa": "ورکشاپ کشف، زبان بک‌لاگ و هم‌راستایی مهندسی با جریان‌های حیاتی درآمد.",
            },
        }
    ]
    db.add(Resume(member_id=sara.id, data_json=json.dumps(sara_resume, ensure_ascii=False)))

    reza = TeamMember(
        slug="reza-karimi",
        name_en="Reza Karimi",
        name_fa="رضا کریمی",
        role_en="Frontend Systems Engineer",
        role_fa="مهندس سیستم‌های فرانت‌اند",
        short_bio_en="Builds calm, high-signal interfaces on top of complex backends — clarity for operators and customers.",
        short_bio_fa="رابط‌های آرام و خوانا روی بک‌اندهای پیچیده — وضوح برای اپراتور و مشتری.",
        sort_order=2,
        is_published=True,
        email="reza@genesis.dev",
        github="https://github.com",
    )
    db.add(reza)
    db.flush()
    reza_resume = empty_resume(reza.name_en, reza.name_fa, reza.role_en, reza.role_fa, reza.email)
    reza_resume["summary"] = {
        "en": "Frontend systems engineer specializing in design systems and operational dashboards for multi-tenant products.",
        "fa": "مهندس فرانت با تمرکز روی دیزاین‌سیستم و داشبوردهای عملیاتی محصولات چندمستاجره.",
    }
    reza_resume["skills"] = ["TypeScript", "React / Next.js", "Design Systems", "Accessibility", "Performance"]
    reza_resume["experiences"] = [
        {
            "title": {"en": "Frontend Engineer", "fa": "مهندس فرانت‌اند"},
            "location": {"en": "Genesis", "fa": "جنسیس"},
            "date": {"en": "2023 — Present", "fa": "۱۴۰۲ — اکنون"},
            "description": {
                "en": "Owns client shells for SaaS consoles, focusing on state clarity and fast paths for power users.",
                "fa": "مالک شِل کلاینت کنسول‌های SaaS با تمرکز روی وضوح state و مسیرهای سریع برای کاربران حرفه‌ای.",
            },
        }
    ]
    db.add(Resume(member_id=reza.id, data_json=json.dumps(reza_resume, ensure_ascii=False)))

    db.commit()

File: src/lib/seed.ts
import { readFileSync, existsSync } from 'fs'
import path from 'path'

import { hashPassword } from '@/lib/auth'
import { prisma } from '@/lib/db'
import { homeContent } from '@/data/homeContent'
import { syncProfileResumes } from '@/lib/profileSync'

export const DEFAULT_HOME = homeContent

export function emptyResume(
  nameEn: string,
  nameFa: string,
  roleEn: string,
  roleFa: string,
  email = '',
): Record<string, unknown> {
  return {
    personal: {
      name: { fa: nameFa || nameEn, en: nameEn },
      tagline: { fa: roleFa || roleEn, en: roleEn },
      avatar: { fa: (nameFa || nameEn).slice(0, 1), en: nameEn.slice(0, 1) },
      location: { fa: 'تهران', en: 'Tehran' },
      phone: { fa: '', en: '' },
      phoneHref: '',
      email,
      badges: [],
    },
    contactsSecondary: [],
    summary: { fa: '', en: '' },
    experiences: [] as Record<string, unknown>[],
    projects: { major: [], small: [] },
    skills: [] as string[],
    education: [],
    certificates: [],
    languages: [],
    references: [],
    footer: {
      text: { fa: 'ISEMPTY', en: 'ISEMPTY' },
      backToTop: { fa: 'بازگشت به بالا', en: 'Back to top' },
    },
  }
}

let seedPromise: Promise<void> | null = null

export async function ensureSeeded() {
  if (!seedPromise) {
    seedPromise = seedIfEmpty().catch((err) => {
      seedPromise = null
      throw err
    })
  }
  await seedPromise
  // Profile JSON is source of truth — keep DB resumes aligned after every API boot path.
  await syncProfileResumes()
}

function loadMahanResumeFromProfile(): Record<string, unknown> | null {
  const profilePath = path.join(process.cwd(), 'src', 'data', 'mahanProfile.json')
  if (!existsSync(profilePath)) return null
  const raw = JSON.parse(readFileSync(profilePath, 'utf-8')) as {
    resume?: Record<string, unknown>
  }
  if (!raw.resume || Object.keys(raw.resume).length === 0) return null
  return raw.resume
}

function buildMahanResume() {
  const fromFile = loadMahanResumeFromProfile()
  if (fromFile) return fromFile

  return emptyResume(
    'Mahan Tahmasbi',
    'ماهان طهماسبی',
    'Backend Developer & System Architect',
    'توسعه‌دهنده بک‌اند و معمار سیستم',
    'mahan.tahmasbi85@gmail.com',
  )
}

const MAHAN_MEMBER = {
  slug: 'mahan-tahmasbi',
  nameEn: 'Mahan Tahmasbi',
  nameFa: 'ماهان طهماسبی',
  roleEn: 'Backend Developer & System Architect',
  roleFa: 'توسعه‌دهنده بک‌اند و معمار سیستم',
  shortBioEn:
    'Backend and systems — scalable APIs, DDD, and taking products from build to market with SEO and tech sales.',
  shortBioFa:
    'بک‌اند و سیستم — API مقیاس‌پذیر، DDD، و رساندن محصول از ساخت تا بازار با سئو و فروش فنی.',
  email: 'mahan.tahmasbi85@gmail.com',
  sortOrder: 1,
  isPublished: true,
} as const

async function seedIfEmpty() {
  const [existingAdmin, members] = await Promise.all([
    prisma.adminUser.findFirst(),
    prisma.teamMember.findMany({ select: { slug: true } }),
  ])
  const existingSlugs = new Set(members.map((m) => m.slug))
  const hasCore =
    existingAdmin &&
    existingSlugs.has('ali-mahdavinia') &&
    existingSlugs.has('mahan-tahmasbi') &&
    existingSlugs.has('reza-karimi')
  if (hasCore) {
    const site = await prisma.siteSettings.findFirst()
    if (site) return
  }

  const username = process.env.ADMIN_USERNAME || 'admin'
  const password = process.env.ADMIN_PASSWORD || 'admin123'
  const passwordHash = await hashPassword(password)

  const contacts = {
    email: 'alimahdavinia125@gmail.com',
    telegram: 'https://t.me/mahdaviniaali',
    linkedin: 'https://www.linkedin.com/in/mahdaviniaali/',
    github: 'https://github.com/mahdaviniaali',
  }

  let aliResume: Record<string, unknown> = {}
  const profilePath = path.join(process.cwd(), 'src', 'data', 'profile.json')
  if (existsSync(profilePath)) {
    const raw = JSON.parse(readFileSync(profilePath, 'utf-8')) as {
      resume?: Record<string, unknown>
      landing?: { contacts?: { type: string; href: string }[] }
    }
    aliResume = raw.resume || {}
    for (const c of raw.landing?.contacts || []) {
      if (c.type in contacts && c.href) {
        ;(contacts as Record<string, string>)[c.type] =
          c.type === 'email' ? c.href.replace('mailto:', '') : c.href
      }
    }
  }

  const emailClean = contacts.email.replace('mailto:', '')

  const mahanResume = buildMahanResume()

  const rezaResume = emptyResume(
    'Reza Karimi',
    'رضا کریمی',
    'Frontend Systems Engineer',
    'مهندس سیستم‌های فرانت‌اند',
    'reza@genesis.dev',
  )
  rezaResume.summary = {
    en: 'Frontend systems engineer specializing in design systems and operational dashboards for multi-tenant products.',
    fa: 'مهندس فرانت با تمرکز روی دیزاین‌سیستم و داشبوردهای عملیاتی محصولات چندمستاجره.',
  }
  rezaResume.skills = ['TypeScript', 'React / Next.js', 'Design Systems', 'Accessibility', 'Performance']
  rezaResume.experiences = [
    {
      title: { en: 'Frontend Engineer', fa: 'مهندس فرانت‌اند' },
      location: { en: 'ISEMPTY', fa: 'ISEMPTY' },
      date: { en: '2023 — Present', fa: '۱۴۰۲ — اکنون' },
      description: {
        en: 'Owns client shells for SaaS consoles, focusing on state clarity and fast paths for power users.',
        fa: 'مالک شِل کلاینت کنسول‌های SaaS با تمرکز روی وضوح state و مسیرهای سریع برای کاربران حرفه‌ای.',
      },
    },
  ]

  await prisma.$transaction(async (tx) => {
    if (!existingAdmin) {
      await tx.adminUser.create({
        data: { username, passwordHash },
      })
    }

    const site = await tx.siteSettings.findFirst()
    if (!site) {
      await tx.siteSettings.create({
        data: {
          brandName: 'ISEMPTY',
          tagline: 'From the void of is empty; We code the light.',
          email: emailClean,
          telegram: contacts.telegram,
          linkedin: contacts.linkedin,
          github: contacts.github,
          homeContent: JSON.stringify(DEFAULT_HOME),
        },
      })
    }

    const existingSlugsInTx = new Set(
      (await tx.teamMember.findMany({ select: { slug: true } })).map((m) => m.slug),
    )

    if (!existingSlugsInTx.has('ali-mahdavinia')) {
      await tx.teamMember.create({
        data: {
          slug: 'ali-mahdavinia',
          nameEn: 'Ali Mahdavinia',
          nameFa: 'علی مهدوی‌نیا',
          roleEn: 'AI Systems & Backend Architect',
          roleFa: 'معمار سیستم‌های AI و بک‌اند',
          shortBioEn:
            'Senior development and technical leadership — enterprise RAG, agents on chatbot.ir, and turning business needs into shippable architecture.',
          shortBioFa:
            'توسعهٔ ارشد و مدیریت فنی — RAG سازمانی و ایجنت‌ها روی chatbot.ir، و تبدیل نیاز کسب‌وکار به معماری قابل تحویل.',
          avatarUrl: '',
          email: emailClean,
          github: contacts.github,
          linkedin: contacts.linkedin,
          telegram: contacts.telegram,
          sortOrder: 0,
          isPublished: true,
          resume: {
            create: {
              dataJson: JSON.stringify(
                Object.keys(aliResume).length
                  ? aliResume
                  : emptyResume(
                      'Ali Mahdavinia',
                      'علی مهدوی‌نیا',
                      'AI Systems & Backend Architect',
                      'معمار سیستم‌های AI و بک‌اند',
                      emailClean,
                    ),
              ),
            },
          },
        },
      })
    }

    if (!existingSlugsInTx.has('mahan-tahmasbi')) {
      await tx.teamMember.create({
        data: {
          ...MAHAN_MEMBER,
          resume: { create: { dataJson: JSON.stringify(mahanResume) } },
        },
      })
    }

    if (!existingSlugsInTx.has('reza-karimi')) {
      await tx.teamMember.create({
        data: {
          slug: 'reza-karimi',
          nameEn: 'Reza Karimi',
          nameFa: 'رضا کریمی',
          roleEn: 'Frontend Systems Engineer',
          roleFa: 'مهندس سیستم‌های فرانت‌اند',
          shortBioEn:
            'Builds calm, high-signal interfaces on top of complex backends — clarity for operators and customers.',
          shortBioFa: 'رابط‌های آرام و خوانا روی بک‌اندهای پیچیده — وضوح برای اپراتور و مشتری.',
          sortOrder: 2,
          isPublished: true,
          email: 'reza@genesis.dev',
          github: 'https://github.com',
          resume: { create: { dataJson: JSON.stringify(rezaResume) } },
        },
      })
    }
  })
}
